src/iterator/mod.rs

```diff
@@ -0,0 +1,307 @@
+//!         signal_hook::SIGINT,
+//!         signal_hook::SIGQUIT,
+//! #       signal_hook::SIGUSR1,
+//!     ])?;
+//! #   // A trick to terminate the example when run as doc-test. Not part of the real code.
+//! #   unsafe { libc::raise(signal_hook::SIGUSR1) };
+//!     'outer: loop {
+//!         // Pick up signals that arrived since last time
+//!         for signal in signals.pending() {
+//!             match signal as libc::c_int {
+//!                 signal_hook::SIGHUP => {
+//!                     // Reload configuration
+//!                     // Reopen the log file
+//!                 }
+//!                 signal_hook::SIGTERM | signal_hook::SIGINT | signal_hook::SIGQUIT => {
+//!                     break 'outer;
+//!                 },
+//! #               signal_hook::SIGUSR1 => return Ok(()),
+//!                 _ => unreachable!(),
+//!             }
+//!         }
+//!         // Do some bit of work ‒ something with upper limit on waiting, so we don't block
+//!         // forever with a SIGTERM already waiting.
+//!     }
+//!     println!("Terminating. Bye bye");
+//!     Ok(())
+//! }
+//! ```
+
+pub mod backend;
+
+use std::borrow::Borrow;
+use std::io::{Error, ErrorKind, Read};
+use std::os::unix::net::UnixStream;
+
+use libc::{self, c_int};
+
+pub use self::backend::Pending;
+use self::backend::PollResult;
+
+/// The main structure of the module, representing interest in some signals.
+///
+/// Unlike the helpers in other modules, this registers the signals when created and unregisters
+/// them on drop. It provides the pending signals during its lifetime, either in batches or as an
+/// infinite iterator.
+///
+/// # Multiple consumers
+///
+/// You may have noticed this structure can be used simultaneously by multiple threads. If it is
+/// done, a signal arrives to one of the threads (on the first come, first serve basis). The signal
+/// is *not* broadcasted to all currently active threads.
+///
+/// A similar thing applies to cloning the structure ‒ at least one of the copies gets the signal,
+/// but it is not broadcasted to all of them.
+///
+/// If you need multiple recipients, you can create multiple independent instances (not by cloning,
+/// but by the constructor).
+///
+/// # Examples
+///
+/// ```rust
+/// # extern crate signal_hook;
+/// #
+/// # use std::io::Error;
+/// # use std::thread;
+/// use signal_hook::iterator::Signals;
+///
+/// #
+/// # fn main() -> Result<(), Error> {
+/// let signals = Signals::new(&[signal_hook::SIGUSR1, signal_hook::SIGUSR2])?;
+/// thread::spawn(move || {
+///     for signal in &signals {
+///         match signal {
+///             signal_hook::SIGUSR1 => {},
+///             signal_hook::SIGUSR2 => {},
+///             _ => unreachable!(),
+///         }
+///     }
+/// });
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug, Clone)]
+pub struct Signals(backend::SignalDelivery<UnixStream, UnixStream>);
+
+type Forever = backend::Forever<UnixStream, UnixStream>;
```

Comment:
Style nitpick: Should they be imported instead of referred as `backend::Something`?

Reply:
This was more or less a left over from the time both were in the same file. I also renamed the Forever struct in the backend because it isn't really iterating for ever by default.